### DISCUSSION TITLE
Ex 2 Task 7

### DISCUSSION THREAD

**Discussion by BenDavisonPetch** [Part-II-Computational-Physics/.github#26] [Q&A] [answered]
I'm currently working on task 7 in exercise 2. The exercise says that the simplest solution would be to take the code from the lectures, and adapt it to use our new Hloc. Upon doing this, my function does not replicate the behaviour of using the full hamiltonian matrix on a state.

```python
def ActWithH(psi, N, h=1.):
    
    hloc = LocalHamiltonian(h).reshape(2,2,2,2)
    d = hloc.shape[0]
    psiOut = np.zeros(psiIn.size)
    for k in range(N - 1):
    # apply local Hamiltonian terms to sites [k,k+1]
        psiOut += np.tensordot(hloc.reshape(d**2, d**2),
                           psiIn.reshape(d**k, d**2, d**(N - 2 - k)),
                           axes=[[1], [1]]).transpose(1, 0, 2).reshape(d**N)

    # apply periodic term
    psiOut += np.tensordot(hloc.reshape(d, d, d, d),
                           psiIn.reshape(d, d**(N - 2), d),
                           axes=[[2, 3], [2, 0]]
                           ).transpose(1, 2, 0).reshape(d**N)

    return psiOut  

# Unit tests
N=3
h=2.
psi = np.array([1,0,0,0,0,0,0,0])
# assert(np.allclose(ActWithH(psi,N,h),CompleteHamiltonian(N,h) @ psi)) # This assert fails
ActWithH(psi,N,h) # output: array([-23., -20., -21., -22., -23., -24., -25., -58.])
CompleteHamiltonian(N,h) @ psi # output: array([-3., -2., -2.,  0., -2.,  0.,  0.,  0.])
```

Could somebody please help me understand what on earth the tensordot is doing and how it works so that I can try and debug it? Without any explanation the code that's given is completely unreadable

**Comment by JNHallen** [comment, marked as answer]
Hi! Your function takes the input "psi", but then you are operating on some other vector "psiIn", which I assume is defined elsewhere in your code.

**Reply by BenDavisonPetch** [reply]
Yep that'll do it - thanks so much, can't believe I didn't catch that

**Comment by aneesa-af** [comment]
I keep getting ValueError: matrix shape must be two-dimensional in reference to ActWithH(psi, N, h=1.) I'm not sure what I'm doing wrong?

**Reply by JNHallen** [reply]
I had a look at your code. The issue is that you have redefined your function LocalHamiltonian so that it returns a sparse matrix (a csr_matrix). You cannot simply reshape a sparse matrix to be four dimensional. Try using a normal array for hloc instead.

**Reply by aneesa-af** [reply]
Is this the same definition of hloc as in task 1?

**Reply by JNHallen** [reply]
Yes it should be.
